src/panel/applets/status/PowerIndicator.vala

```diff
@@ -146,11 +146,130 @@ public class BatteryIcon : Gtk.Box {
 	}
 }
 
+const string POWER_PROFILES_DBUS_NAME = "net.hadess.PowerProfiles";
+const string POWER_PROFILES_DBUS_OBJECT_NAME = "/net/hadess/PowerProfiles";
+
+namespace net.hadess {
+	[DBus (name = "net.hadess.PowerProfiles")]
+	public interface PowerProfilesDBus : Object {
+		[DBus (name = "Profiles")]
+		public abstract HashTable<string, Variant>[] profiles { owned get; }
+
+		[DBus (name = "ActiveProfile")]
+		public abstract string active_profile { owned get; set; }
+	}
+}
+
+public class PowerProfilesSelector : Gtk.Box {
+	// Power profile toggles
+	private Gtk.RadioButton radio_power_save;
+	private Gtk.RadioButton radio_power_balanced;
+	private Gtk.RadioButton radio_power_performance;
+
+	// these are used to temporarily block signals and prevent an infinite loop
+	private ulong radio_power_save_id;
+	private ulong radio_power_balanced_id;
+	private ulong radio_power_performance_id;
+
+	public PowerProfilesSelector(net.hadess.PowerProfilesDBus profiles_proxy) {
+		orientation = Gtk.Orientation.VERTICAL;
+		
+		var profiles = new GenericSet<string>(str_hash, str_equal);
+
+		foreach (HashTable<string, Variant> profile in profiles_proxy.profiles) {
+			var profile_value = profile.get("Profile");
+
+			if (!profile_value.is_of_type(VariantType.STRING)) {
+				warning("unexpected type for power profile: is not of VariantType.STRING");
+				return;
+			}
+			
+			profiles.add(profile_value.get_string());
+		}
+		
+		if (profiles.length < 2) {
+			// need at least two options for it to be meaningful
+			warning("Fewer than two power profile options");
+			return;
+		}
+
+		// initialize state
+		on_active_profile_changed(profiles_proxy.active_profile);
+
+		DBusProxy? proxy = profiles_proxy as DBusProxy;
+		proxy.g_properties_changed.connect(() => {
+			on_active_profile_changed(profiles_proxy.active_profile);
+		});
+
+		var sep = new Gtk.Separator(Gtk.Orientation.HORIZONTAL);
+		pack_start(sep, false, false, 1);
+
+		var header = new Gtk.Label("");
+		header.set_markup("<b>%s</b>".printf(_("Power Mode")));
+		header.set_halign(Gtk.Align.CENTER);
+		pack_start(header);
+
+		var power_profiles_radio_box = new Gtk.Box(Gtk.Orientation.HORIZONTAL, 1);
+
+		if (profiles.contains("power-saver")) {
+			radio_power_save = new Gtk.RadioButton.with_label_from_widget(null, _("Power Saver"));
+			radio_power_save_id = radio_power_save.toggled.connect(() => {
+				profiles_proxy.active_profile = "power-saver";
+			});
```

Comment:
`toggled` is triggered both when the radio button is activated *and* deactivated, so be sure to check the current button state:

```suggestion
			radio_power_save_id = radio_power_save.toggled.connect(() => {
				if (radio_power_save.is_active()) {
					profiles_proxy.active_profile = "power-saver";
				}
			});
```

Reply:
the handlers for all three are blocked and then unblocked in `on_active_profile_changed()`, so that shouldn't be necessary